Type errors report wrong source location (often line 1)

## Description

Some type errors report the wrong source location, often pointing to line 1 instead of the actual error location.

## Reproduction

```ez
import @std
using std

do main() {
    if nil {  // Error should point here (line 5)
        println("truthy")
    }
}
```

## Error

```
error[E3001]: if condition must be bool, got nil
  --> main.ez:1:1
   |
 1 | import @std
   | ^ types do not match
```

## Expected Behavior

Error should point to line 5 where `if nil` appears.

## Actual Behavior

Error points to line 1, making it difficult to locate the actual problem.

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -1814,6 +1814,18 @@ func (tc *TypeChecker) getExpressionPosition(expr ast.Expression) (int, int) {
 		return e.Token.Line, e.Token.Column
 	case *ast.MemberExpression:
 		return e.Token.Line, e.Token.Column
+	case *ast.NilValue:
+		return e.Token.Line, e.Token.Column
+	case *ast.MapValue:
+		return e.Token.Line, e.Token.Column
+	case *ast.StructValue:
+		return e.Token.Line, e.Token.Column
+	case *ast.NewExpression:
+		return e.Token.Line, e.Token.Column
+	case *ast.PostfixExpression:
+		return e.Token.Line, e.Token.Column
+	case *ast.RangeExpression:
+		return e.Token.Line, e.Token.Column
 	default:
 		return 1, 1
 	}

--- a/pkg/typechecker/typechecker_test.go
+++ b/pkg/typechecker/typechecker_test.go
@@ -1576,3 +1576,61 @@ do main() {
 	tc := typecheck(t, input)
 	assertNoErrors(t, tc)
 }
+
+// ============================================================================
+// Error Location Tests (Fixes #345)
+// ============================================================================
+
+func TestTypeErrorLocationNilInIfCondition(t *testing.T) {
+	// Test that type errors report correct source location
+	// Fixes #345
+	input := `do main() {
+	if nil {
+		temp x int = 1
+	}
+}`
+	tc := typecheck(t, input)
+
+	if !tc.Errors().HasErrors() {
+		t.Fatal("expected type error for nil in if condition")
+	}
+
+	// Check that error is NOT at line 1
+	errs := tc.Errors()
+	for _, err := range errs.Errors {
+		if err.ErrorCode == errors.E3001 {
+			if err.Line == 1 {
+				t.Errorf("error should not be at line 1, got line %d", err.Line)
+			}
+			if err.Line != 2 {
+				t.Errorf("error should be at line 2, got line %d", err.Line)
+			}
+			return
+		}
+	}
+	t.Error("expected error E3001 not found")
+}
+
+func TestTypeErrorLocationMapValue(t *testing.T) {
+	// Test that map value errors report correct location
+	input := `do main() {
+	if {"a": 1} {
+		temp x int = 1
+	}
+}`
+	tc := typecheck(t, input)
+
+	if !tc.Errors().HasErrors() {
+		t.Fatal("expected type error for map in if condition")
+	}
+
+	errs := tc.Errors()
+	for _, err := range errs.Errors {
+		if err.ErrorCode == errors.E3001 {
+			if err.Line == 1 {
+				t.Errorf("error should not be at line 1, got line %d", err.Line)
+			}
+			return
+		}
+	}
+}

